bf2fc6cc711aee1a0c2a/kas-fleet-manager
# feat: Expose GCP configuration

## Description

Expose configurability of GCP credentials as a secret file.
Related to https://issues.redhat.com/browse/MGDSTRM-9132.

Information about GCP API credentials can be found in: https://cloud.google.com/iam/docs/creating-managing-service-account-keys

* For local development with a local process a file named `gcp.api-credentials` is expected in the secrets directory
* For development in an OSD cluster it is created as the `gcp.api-credentials` in the K8s secret named 'kas-fleet-manager' from the secrets template
  * Its value can be configured through the `GCP_API_CREDENTIALS` template parameter
    * The `GCP_API_CREDENTIALS` template parameter must be a base64 encoded string of the GCP API credentials configuration file contents (similarly to the image-pull.dockerconfigjson secret)
* The default location of the credentials file can be modified by using the newly added flag '--gcp-api-credentials-file' which by default it is set to 'secrets/gcp.api-credentials'

Additional remarks:
* Validation of the credentials file contents is only performed when `gcp` is listed as a supported cloud provider in the providers configuration file
* The credentials file is not mandatory when `gcp` is not a supported cloud provider
* If the credential file exists but it is empty or only contains blank it is treated in the same way as if it did not exist
* To perform validation at Go level a new library https://github.com/go-playground/validator has been used. I think this might be useful to perform validations at a general level

This change shouldn't cause any issue when merging it as `gcp` is not a supported provider at the moment in stage nor in production. To enable it in those environments it will need to be added as a supported provider and the corresponding secrets configured accordingly with the secrets mechanism used in stage and production.


## Verification Steps

* Verify the different scenarios and combinations between having the GCP credentails file and GCP as a supported provider.
* Verify that KAS Fleet Manager runs successfully both as a local process and in an OSD dev cluster, with and without GCP credentials and with and without gcp as a supported cloud provider.

Credentials can be obtained from Vault. Ask them to me for directions on how to get them if needed.



## Checklist (Definition of Done)
